### PR TITLE
Bump minimum TLS protocol to TLSv1.2

### PR DESCRIPTION
**Pull request by stevendanna**
Some environments are subject to strict rules about the permitted TLS
protocol verion and available ciphers. This change makes the former
configurable for both HTTPS and gRPC.

The gRPC protocol rejects anything under TLSv1.2; however, it is still
useful to configure the tls protocol version explicitly because many
automated scanners do not recognize gRPC.

Signed-off-by: Steven Danna <steve@chef.io>

### CODE REVIEW

**Comment by ericchiang** [comment]
Instead of making this configurable, I'm fine enforcing TLS 1.2 and up

**Comment by stevendanna** [comment]
> Instead of making this configurable, I'm fine enforcing TLS 1.2 and up

That would also meet my personal needs.  Happy to modify this PR to do that if y'all are OK with the reduced client compatibility.

**Comment by ericchiang** [comment]
https://www.zdnet.com/article/chrome-edge-ie-firefox-and-safari-to-disable-tls-1-0-and-tls-1-1-in-2020/

TLS 1.0 and 1.1 are being phased out. I don't feel that dex needs to support them, particularly for any protocol like gRPC. Backend LDAP servers might be the one exception here, since some of that tech tends to be older.

Let's enforce TLS 1.2 for now.

**Comment by ericchiang** [comment]
There's a pretty big diff here. I'd only expect the tls.Config to change.

**Comment by stevendanna** [comment]
> There's a pretty big diff here. I'd only expect the tls.Config to change.

Most of the changes are moving code around such that I could make sure we are using the tls.Config without too much duplication across different code branches. In the old code, when not using a ClientCA we didn't generate our own tls.Config.

https://github.com/dexidp/dex/blob/master/cmd/dex/serve.go#L147

The majority of the diff is a result of changing it so that we are using a tls.Config in that case.

If you'd prefer, I can make the diff smaller, the result will just be an opportunity for the tls.Config to drift between those two branches.

File: cmd/dex/serve.go [hunk 1]
```diff
@@ -133,23 +139,17 @@ func serve(cmd *cobra.Command, args []string) error {
...
 
-			tlsConfig := tls.Config{
-				Certificates: []tls.Certificate{cert},
-				ClientAuth:   tls.RequireAndVerifyClientCert,
-				ClientCAs:    cPool,
-			}
+			tlsConfig.ClientAuth = tls.RequireAndVerifyClientCert
+			tlsConfig.ClientCAs = cPool
+
+			// Only add metrics if client auth is enabled
```
[the remark this answers is not in the document]
Comment (srenatus) [reply cmd/dex/serve.go]: 🤔 Interesting. Might be a useful follow-up PR to move that out of this `if` branch. Unless someone remembers the rationale for this...

**Comment by srenatus** [comment]
Thank you! I'll merge this. A potential follow-up would be to have gRPC and HTTPS share a `*tls.Config` -- at least in our deployment, external (HTTP) clients talk to a reverse proxy anyways, which allows connection with no client cert; but internal comms (including the proxy -> dex) need mutual TLS. With that change, we could enforce that only authenticated clients (i.e., the proxy) talk to dex's HTTPS endpoint. But anyways, that's just a potential follow-up. 😄

**Pull request merged**